src/ringct/rctTypes.h

```diff
@@ -221,9 +228,9 @@ namespace rct {
       RCTTypeNull = 0,
       RCTTypeFull = 1,
       RCTTypeSimple = 2,
-      RCTTypeFullBulletproof = 3,
-      RCTTypeSimpleBulletproof = 4,
+      RCTTypeBulletproof = 3,
     };
+    enum RangeProofType { RangeProofBorromean, RangeProofBulletproof, RangeProofMultiOutputBulletproof, RangeProofPaddedBulletproof };
```

Comment:
I'm still confused about the role of these different types of BPs:

1. Looking at wallet2.cpp, `RangeProofBulletproof` seems to be used only when the tx has one output. However, basically all txes are supposed to have at least 2 outputs (one dummy in the case of sweep txes).

2. `RangeProofMultiOutputBulletproof` seems never used in the actual tx construction (never used in wallet2). It seems like a generalization of `RangeProofPaddedBulletproof` where more than 16 outputs can be included in a tx (as opposed to `RangeProofPaddedBulletproof` which is limited to just one `Bulletproof` and thus up to 16 outputs).

As such, it appears that just the `RangeProofMultiOutputBulletproof` type suffices. Am I missing something?

Reply:
RangeProofBulletproof is the existing one, with one proof per amount. RangeProofMultiOutputBulletproof is the "optimal size" one which creates, eg, a proof for 1 amount and a proof for 4 amounts when asked to prove 5 amounts. RangeProofPaddedBulletproof is the new one which will prove 8 amounts (next power of 2) when asked to prove 5.RangeProofMultiOutputBulletproof would generate proofs that do not comply with ArticMine's recommendation.

Reply:
Why does the `gen_bp_tx_invalid_1_1` core test fail then? I expected it to create a tx with two `Bulletproof`s each having one element in `V`.

Reply:
Because ArticMine decided it was better that way. You can ask him for the details of the reasoning.

Reply:
I see now why `gen_bp_tx_invalid_1_1` is failing; the `expand_transaction_1` function simply rejects multiple bulletproofs:
```diff
-           if (rv.p.bulletproofs.size() != tx.vout.size())
+           if (rv.p.bulletproofs.size() != 1)
```

In your above explanation,

> `RangeProofBulletproof` is the existing one, with one proof per amount.

you mean the current testnet by the "existing one", I suppose? I don't see any point in caring about the compatibility and in preserving the old format. We can simply throw away the old/obsolete format and roll back the testnet, no?

> `RangeProofMultiOutputBulletproof` would generate proofs that do not comply with ArticMine's recommendation.

So in the end, `RangeProofPaddedBulletproof` is the only usable BP type. Why bother with the other two unused types? Removing them seems to help simplify the code.

Reply:
The current one is the one on current testnet, yes.Those extra variants don't use up much code, and we might use them again in the future as dynamics change, so I don't think it's necessarily best to remove them. If nothing else, it allows one to test they're rejected :P If you feel strongly about it though I'll remove them. Testnet will be rolled back, so no backward compatibility to worry about.

Reply:
OK, fine with keeping them, but at least I want `RangeProofBulletproof` to be renamed to something else that carries some information. Maybe something like `RangeProofNonpaddedBulletproof`? The current form `RangeProofBulletproof` seems too generic.